resource/wrapper/retryresource/simple_resource.go

```diff
@@ -86,11 +92,6 @@ func (r *resourceWrapper) EnsureDeleted(ctx context.Context, obj interface{}) er
 	return nil
 }
 
-func (r *resourceWrapper) Name() string {
+func (r *simpleResource) Name() string {
 	return r.resource.Name()
 }
-
-// Wrapped implements internal.Wrapper interface.
-func (r *resourceWrapper) Wrapped() resource.Interface {
```

Comment:
I'm uncertain how `internal.Underlying` is supposed to work now when you removed that. I also don't see this method in `crudResource` type.

Reply:
I think this underlying thing is not even necessary. I do not see what it is supposed to do now that the implementation is heavily simplified. When you check `retry_resource.go` you see that we get the underlying resource only from the injected resource, which is the resource we want to wrap. `internal.Underlying` only ever returns the injected resource as this is what we want to wrap. We do not wrap wrapped resources, don't we? I don't really follow the intentions of this unwrapping really. 

Let's assume we have an operator and it wants to wrap its resource with metrics and retries. So we wrap resource with metrics and then wrap metrics with retry. On error retry executes metrics again which executes resource. This is all simple and fine. What do I miss?

Reply:
Let's say you want to wrap the resources with metrics and they with retires.

So `metricsresource.New` here will return `*Resource` that implements `resource.Interface`. Then next `crud.Interface` type assertion in `retryresource.New` won't work because `crud.Interface` is just a field in the returned type. That's why we need unwrapping.

Reply:
When A is the resource we want to wrap with B which in turn is wrapped by C the execution is C, B, A. Calling `C.Name()` dispatches all the way through and returns `A.Name()`. I think we do not need any unwrapping at all as calling calls what we want in the order it is wrapped. I will do some testing when I find time.

Reply:
How `C.Name()` is dispatched is irrelevant.

The problem is that this https://github.com/giantswarm/operatorkit/pull/316/files#diff-d2ff54b878a938f2f2d9df63065b97eaR20 doesn't implement `crud.Interface` even if it wraps `crud.Interface`. Let's assume `B` is of type `retryresource.Resource` of this implementation:

- `A.(crud.Interface)` will pass.
- `A` gets wrapped with `B`
- `B` calls `crud.NewResource` which returns `resource.Interface` and not `crud.Interface`
- `B.(crud.Interface)` will not pass.
- `C` will wrap only `resource.Interface` methods of `B` (including mentioned `Name` method) instead of `crud.Interace` methods of `B` internals. 

I'm playing around to see if this can be overcame by exposing `CRUD()` method on `*crud.Resoruce` and wrapping friends.

Reply:
It's even worse than I thought. Code below prints `ok = false` when executed. Meaning that even `B` won't wrap `crud.Interface` methods.

```go
package main

import (
	"context"
	"fmt"
	"log"

	"github.com/giantswarm/micrologger/microloggertest"

	"github.com/giantswarm/operatorkit/resource"
	"github.com/giantswarm/operatorkit/resource/crud"
)

func main() {
	var err error

	var r resource.Interface
	{
		c := crud.ResourceConfig{
			CRUD:   &myCRUD{},
			Logger: microloggertest.New(),
		}
		r, err = crud.NewResource(c)
		if err != nil {
			log.Fatal(err)
		}

	}

	_, ok := r.(crud.Interface)
	fmt.Printf("ok = %v\n", ok)
}

type myCRUD struct {
}

func (*myCRUD) Name() string {
	return "myCRUD"
}
func (*myCRUD) GetCurrentState(ctx context.Context, obj interface{}) (interface{}, error) {
	return nil, nil
}
func (*myCRUD) GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error) {
	return nil, nil
}
func (*myCRUD) NewUpdatePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*crud.Patch, error) {
	return nil, nil
}
func (*myCRUD) NewDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*crud.Patch, error) {
	return nil, nil
}
func (*myCRUD) ApplyCreateChange(ctx context.Context, obj, createChange interface{}) error {
	return nil
}
func (*myCRUD) ApplyDeleteChange(ctx context.Context, obj, deleteChange interface{}) error {
	return nil
}
func (*myCRUD) ApplyUpdateChange(ctx context.Context, obj, updateChange interface{}) error {
	return nil
}
```

Reply:
This should fix the issue: https://github.com/giantswarm/operatorkit/pull/318
It also tests it. With this approach we can remove the old unwrapping logic. But it's at expense of having `CRUD` method in `*crud.Resource` (not exposed in the interface).